### PR TITLE
Add example for 3D adaptive histogram equalization (AHE)

### PR DESCRIPTION
**Pull request by m-albert**
## Description

New gallery example illustrating functionality added in this PR https://github.com/scikit-image/scikit-image/pull/4598.

The example compares the results of applying AHE to a 3D image and a degraded version of it to illustrate how AHE can be used to recover contrast and normalize images.

image

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by emmanuelle** [comment]
Hi @m-albert thank you for the PR! The Travis CI is currently failing for the builds using minimal requirements, I believe that it's related to matplotlib. Could you please take a look? Thanks :-).

**Comment by m-albert** [comment]
Hey @emmanuelle, thanks for making me aware of it :) Should be fixed.

**Comment by alexdesiqueira** [comment]
Hey @m-albert, thank you very much for this — it looks great!
I have a suggestion, though. Could we use an _actual_ image, or a more well-behaved set, as an example? For me it's kinda hard to see the AHE effects on this set.
Thanks again!

**Comment by m-albert** [comment]
Hey @alexdesiqueira, thanks for your suggestion!

I do agree that the image is not super clear in the sense that the viewer does not immediately recognize much in it. However it could exemplify e.g. a dense measurement of something in 3d. And I thought including the degradation along x would help the illustration...

Would you have a suggestion for a better 3D image? Or could you point me to available example volumes within scikit-image?

Alternatively, I can think of these ways of making the illustration clearer:
1) including non-adaptive histogram equalization for comparison and/or
2) choosing a different random seed.

That could look something like this:

image

**Comment by alexdesiqueira** [comment]
@m-albert thank you, that looks great already! More context surely made it better.
I was thinking on an image in the lines of [cells.tif](https://github.com/scikit-image/skimage-tutorials/blob/master/images/cells.tif), which we use on skimage 3D tutorials. It's a large image, though; maybe a slice of it would suffice... what do you think, @emmanuelle?
Thanks again!

**Comment by jni** [comment]
@m-albert @alexdesiqueira I actually disagree with Alex here, I think this example shows very nicely how AHE is able to give you a similar image for both degraded and non-degraded images, starting with a *very* degraded image. If we add cells.tif, I would suggest using it *in addition to* the synthetic example.

On the other hand, we could also use cells.tif together with some synthetic degradation...

**Comment by alexdesiqueira** [comment]
@m-albert never mind me then; @jni's idea is better!

**Comment by m-albert** [comment]
@alexdesiqueira I wasn't aware of the cell dataset, I think that's a great suggestion!

> On the other hand, we could also use cells.tif together with some synthetic degradation...


@jni Great idea, here's a combination of the cells with synthetic degradation. What do you think guys?

image

**Comment by alexdesiqueira** [comment]
@m-albert this looks amazing! Thank you!

**Comment by jni** [comment]
> What do you think guys?

:heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes: :heart_eyes:

**Comment by m-albert** [comment]
I'm just realizing that if skimage-tutorials can be used for stably hosting images I could have uploaded a zebrafish embryo like one of these ones: https://twitter.com/i/status/1259893967681916930. Maybe for the next example :)

**Comment by m-albert** [comment]
Hmm some of the checks fail but I don't understand why. The output appears unrelated to my added file. What can I do? Thanks for any help :)

```
=========================== short test summary info ===========================
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::TestColorconv::test_lab_full_gamut
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::TestColorconv::test_rgb2gray_alpha
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape0-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape2-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape3-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/color/tests/test_colorconv.py::test_nD_gray_conversion[shape4-rgb2gray]
FAILED python37-x64/lib/site-packages/skimage/feature/tests/test_corner.py::test_shape_index
FAILED python37-x64/lib/site-packages/skimage/filters/tests/test_gaussian.py::test_multichannel
FAILED python37-x64/lib/site-packages/skimage/filters/tests/test_ridges.py::test_border_warning[sato]
FAILED python37-x64/lib/site-packages/skimage/filters/tests/test_ridges.py::test_border_warning[hessian]
FAILED python37-x64/lib/site-packages/skimage/io/tests/test_pil.py::test_all_color
FAILED python37-x64/lib/site-packages/skimage/io/tests/test_pil.py::test_all_mono
FAILED python37-x64/lib/site-packages/skimage/restoration/tests/test_denoise.py::test_denoise_bilateral_nan
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_bf
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_cg
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_cg_mg
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_cg_j
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_types
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_reorder_labels
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_2d_inactive
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_3d
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_3d_inactive
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_multispectral_2d
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_multispectral_3d
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_spacing_0
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_spacing_1
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_trivial_cases
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_length2_spacing
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_isolated_seeds
FAILED python37-x64/lib/site-packages/skimage/segmentation/tests/test_random_walker.py::test_prob_tol
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_radon_transform.py::test_radon_circle
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_warps.py::test_swirl
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_warps.py::test_downsize_anti_aliasing_invalid_stddev
FAILED python37-x64/lib/site-packages/skimage/transform/tests/test_warps.py::test_boll_array_warnings
== 34 failed, 5828 passed, 163 skipped, 18289 warnings in 130.10s (0:02:10) ===
Command exited with code 1
```

**Comment by rfezzani** [comment]
These CI failures are fixed by #4731 and #4735, you need to rebase if you want an all green CI ;)

**Comment by m-albert** [comment]
Oh I see, thanks a lot @rfezzani :)

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+plt.subplots_adjust(left=0.,
+                    bottom=-.05,
+                    right=1.03,
+                    top=1.19,
+                    wspace=-0.2,
+                    hspace=-0.45)
+
+# highlight AHE
+rect_ax = fig.add_axes([0, 0, 1, 1], facecolor='none')
+rect_ax.axis(False)
```
Comment (rfezzani) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:214]:
```suggestion
rect_ax.set_axis_off()
```
This line is causing CI failures ;)

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import matplotlib.patches as patches
+from matplotlib import cm, colors
+from mpl_toolkits.mplot3d import Axes3D
+
+import numpy as np
+from skimage import exposure
+from skimage import io
```
Comment (rfezzani) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:23]: Please use `imageio` instead.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+
+#############
+# Prepare data and apply hist. eq.
+#############
+
+im_orig = io.imread('https://github.com/scikit-image/'
+                    'skimage-tutorials/blob/master/images/cells.tif?raw=True',
+                    plugin='tifffile')
```
Comment (rfezzani) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:32]:
```suggestion
im_orig = imageio.volread('https://github.com/scikit-image/'
                    'skimage-tutorials/blob/master/images/cells.tif?raw=True')
```

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py (hunk 1 above)
Comment (m-albert) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:214]: Thanks!

**Comment by m-albert** [comment]
PR should be ready from my side. Thanks a lot @rfezzani for reviewing!

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+import imageio
+
+
+#############
+# Prepare data and apply hist. eq.
+#############
+
+im_orig = imageio.volread('https://github.com/scikit-image/'
+                          'skimage-tutorials/blob/master/'
+                          'images/cells.tif?raw=True')
```
Comment (emmanuelle) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:31]:
```suggestion
from skimage.data import image_fetcher
path = image_fetcher.fetch('data/cells.tif')
img_orig = io.imread(path) 
```

Comment (emmanuelle) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:31]: this way we can use our new fetching mechanism with pooch, and if the data have already been downloaded once, they'll be cached for the next time they are opened.

**Comment by emmanuelle** [comment]
I just left a comment about I/O, if you agree with it @m-albert please accept it and then I'll merge. Thanks for a great example!

**Comment by emmanuelle** [comment]
You need to import `io` in the gallery example (see CI failure).

**Comment by m-albert** [comment]
Thanks for your review @emmanuelle! I incorporated your suggestion and fixed some `matplotlib` issues that started to show up with its 3.3 release that came out in the meantime.

And I think the tests are as green as I can get them after rebasing :)

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    rect_shape = np.array(im_orig.shape) + 2
+    for line in lines:
+        ax.plot((line * rect_shape)[:, 0] - 1,
+                (line * rect_shape)[:, 1] - 1,
+                (line * rect_shape)[:, 2] - 1,
+                linewidth=1, color='grey')
+
+# add boxes illustrating the kernels
+ns = np.array(im_orig.shape) // kernel_size - 1
+for iax, ind, rect_shape in zip([1] + [2] * 4,
```
Comment (soupault) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:151]: `iax, ind` could be renamed to improve readibility.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    # matplotlib/matplotlib/issues/17172#issuecomment-617546105
+    xyzlim = np.array([ax.get_xlim3d(),
+                       ax.get_ylim3d(),
+                       ax.get_zlim3d()]).T
+    XYZlim = np.asarray([min(xyzlim[0]), max(xyzlim[1])])
+    ax.set_xlim3d(XYZlim)
+    ax.set_ylim3d(XYZlim)
+    ax.set_zlim3d(XYZlim * 0.55)
+
+    try:
```
Comment (soupault) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:192]: Is this neccessary?

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    ax.set_ylabel('y', labelpad=-25)
+    ax.text2D(0.63, 0.2, "z", transform=ax.transAxes)
+    ax.set_xticks([])
+    ax.set_yticks([])
+    ax.set_zticks([])
+    ax.grid(False)
+    ax.elev = 30
+
+
+plt.subplots_adjust(left=0.01,
```
Comment (soupault) [inline doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:207]: Can be shortened to 1-2 lines.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+#############
+# Prepare data and apply hist. eq.
+#############
+
+from skimage.data import image_fetcher
+path = image_fetcher.fetch('data/cells.tif')
+im_orig = io.volread(path)
+im_orig = im_orig.transpose()
+
+# apply some rescaling
```
[the remark this answers is not in the document]
Comment (soupault) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:34]: `some` is somewhat not informative. I would change to `Apply rescaling`, `Rescale image intensity from 0 to 1`, etc.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+from mpl_toolkits.mplot3d import Axes3D
+
+import numpy as np
+from skimage import exposure
+import imageio as io
+
+
+#############
+# Prepare data and apply hist. eq.
+#############
```
[the remark this answers is not in the document]
Comment (m-albert) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:27]: True, the hashtags can be misunderstood for sphinx formatting

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (m-albert) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:34]: Good point.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+                                 kernel_size=kernel_size,
+                                 clip_limit=clip_limit)
+     for im in [im_orig, im_degraded]]
+
+
+#############
+# define functions to help plot the data
+#############
+
+def get_rgba(scalars, cmap, vmin=0., vmax=1., alpha=0.2):
```
[the remark this answers is not in the document]
Comment (m-albert) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:68]: Definitely.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py (hunk 5 above)
Comment (m-albert) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:151]: Renamed to `axis_ind, vertex_ind, box_shape`, hope that's clearer.

File: doc/examples/color_exposure/plot_adapt_hist_eq_3d.py (hunk 6 above)
Comment (m-albert) [reply doc/examples/color_exposure/plot_adapt_hist_eq_3d.py:192]: I think so, see this thread (commented in the code) https://github.com/matplotlib/matplotlib/issues/17172#issuecomment-617546105. Definitely `napari` will be better for 3D plotting!

**Comment by m-albert** [comment]
Hey @soupault, thanks a lot for your review and feedback! I incorporated all of your suggestions for the code.

> For the figure, I think it would be clearer if the captions from the left edge are put above the subfigures. Additionally, the captions for the bottom right subfigures would be helpful to better differentate the first row from the second. Summarized:

Thanks for the detailed cosmetic suggestions :) I played around a little with caption rearrangements and think that
- for now I prefer the tabular format (at least I couldn't come up with individual captions that I'd feel would increase readability)
- as you suggest, stating more clearly what the input is and how it becomes modified definitely improves the figure (that might also differentiate a bit more the first and second rows).

Let me know what you think!

In unrelated matters, something must have changed in `matplotlib` 3.3. as I'm getting changes in subfigure arrangement and some 3d properties.. hope that'll stay relatively stable in the near future.

image

**Comment by rfezzani** [comment]
I agree with @soupault, what a piece of work @m-albert !! Thank you very much for your amazing contribution :clap:

**Comment by m-albert** [comment]
Thanks for all the feedback, encouragement and great vibes guys :)

**Comment by jni** [comment]
I thought this had merged ages ago! Oops! 😅 Better late than never and thank you yet again @m-albert! 🎉🎉

**Pull request merged**